File: chapters/chp5_code/vec_field.rb
#!/usr/bin/env ruby

class VectorField2D
  def initialize(p_func, q_func)
    @P = p_func
    @Q = q_func
    @h = 1e-5
  end
  
  def evaluate(x, y)
    [@P.call(x, y), @Q.call(x, y)]
  end
  
  def divergence(x, y)
    dP_dx = (@P.call(x + @h, y) - @P.call(x - @h, y)) / (2 * @h)
    dQ_dy = (@Q.call(x, y + @h) - @Q.call(x, y - @h)) / (2 * @h)
    dP_dx + dQ_dy
  end
  
  def curl(x, y)
    dQ_dx = (@Q.call(x + @h, y) - @Q.call(x - @h, y)) / (2 * @h)
    dP_dy = (@P.call(x, y + @h) - @P.call(x, y - @h)) / (2 * @h)
    dQ_dx - dP_dy
  end
  
  def save_field_data(x_range, y_range, resolution, filename)
    File.open(filename, 'w') do |file|
      file.puts "x,y,P,Q,magnitude,divergence,curl"
      
      x_min, x_max = x_range
      y_min, y_max = y_range
      
      dx = (x_max - x_min) / resolution.to_f
      dy = (y_max - y_min) / resolution.to_f
      
      (0..resolution).each do |i|
        (0..resolution).each do |j|
          x = x_min + i * dx
          y = y_min + j * dy
          
          field = evaluate(x, y)
          magnitude = Math.sqrt(field[0]**2 + field[1]**2)
          div = divergence(x, y)
          curl_val = curl(x, y)
          
          file.puts "#{x},#{y},#{field[0]},#{field[1]},#{magnitude},#{div},#{curl_val}"
        end
      end
    end
    
    puts "Vector field data written to #{filename}"
  end
end

puts "2D Vector Field Analysis (Ruby version)"
puts "======================================="

# Example 1: Circulation field F = <-y, x>
circulation_field = VectorField2D.new(
  ->(x, y) { -y },
  ->(x, y) { x }
)

puts "Example 1: Circulation field F = <-y, x>"
circulation_field.save_field_data([-10, 10], [-10, 10], 25, "circulation_field_ruby.csv")

# Example 2: Radial field F = <x, y>
radial_field = VectorField2D.new(
  ->(x, y) { x },
  ->(x, y) { y }
)

puts "Example 2: Radial field F = <x, y>"
radial_field.save_field_data([-5, 5], [-5, 5], 25, "radial_field_ruby.csv")

File: chapters/chp5_code/vec_field.c
#include <stdio.h>
#include <math.h>

double P_circulation(double x, double y) { return -y; }
double Q_circulation(double x, double y) { return x; }

double P_radial(double x, double y) { return x; }
double Q_radial(double x, double y) { return y; }

double divergence(double (*P)(double, double), double (*Q)(double, double),
                 double x, double y, double h) {
    double dP_dx = (P(x + h, y) - P(x - h, y)) / (2 * h);
    double dQ_dy = (Q(x, y + h) - Q(x, y - h)) / (2 * h);
    return dP_dx + dQ_dy;
}

double curl_2d(double (*P)(double, double), double (*Q)(double, double),
               double x, double y, double h) {
    double dQ_dx = (Q(x + h, y) - Q(x - h, y)) / (2 * h);
    double dP_dy = (P(x, y + h) - P(x, y - h)) / (2 * h);
    return dQ_dx - dP_dy;
}

void save_field_data(double (*P)(double, double), double (*Q)(double, double),
                    double x_min, double x_max, double y_min, double y_max,
                    int resolution, const char* filename) {
    FILE* file = fopen(filename, "w");
    if (!file) return;
    
    fprintf(file, "x,y,P,Q,magnitude,divergence,curl\n");
    
    double dx = (x_max - x_min) / resolution;
    double dy = (y_max - y_min) / resolution;
    double h = 1e-5;
    
    for (int i = 0; i <= resolution; i++) {
        for (int j = 0; j <= resolution; j++) {
            double x = x_min + i * dx;
            double y = y_min + j * dy;
            
            double p_val = P(x, y);
            double q_val = Q(x, y);
            double magnitude = sqrt(p_val * p_val + q_val * q_val);
            double div = divergence(P, Q, x, y, h);
            double curl_val = curl_2d(P, Q, x, y, h);
            
            fprintf(file, "%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f\n",
                    x, y, p_val, q_val, magnitude, div, curl_val);
        }
    }
    
    fclose(file);
    printf("Vector field data written to %s\n", filename);
}

int main() {
    printf("2D Vector Field Analysis (C version)\n");
    printf("====================================\n");
    
    printf("Example 1: Circulation field F = <-y, x>\n");
    save_field_data(P_circulation, Q_circulation, -10, 10, -10, 10, 25, "circulation_field_c.csv");
    
    printf("Example 2: Radial field F = <x, y>\n");
    save_field_data(P_radial, Q_radial, -5, 5, -5, 5, 25, "radial_field_c.csv");
    
    return 0;
}